### PR TITLE
[SPARK-44479][PYTHON] Fix ArrowStreamPandasUDFSerializer to accept no-column pandas DataFrame

### PR DESCRIPTION
**Pull request by ueshin**
### What changes were proposed in this pull request?

Fixes `ArrowStreamPandasUDFSerializer` to accept no-column pandas DataFrame.

```py
>>> def _scalar_f(id):
...   return pd.DataFrame(index=id)
...
>>> scalar_f = pandas_udf(_scalar_f, returnType=StructType())
>>> df = spark.range(3).withColumn("f", scalar_f(col("id")))
>>> df.printSchema()
root
 |-- id: long (nullable = false)
 |-- f: struct (nullable = true)

>>> df.show()
+---+---+
| id|  f|
+---+---+
|  0| {}|
|  1| {}|
|  2| {}|
+---+---+
```

### Why are the changes needed?

The above query fails with the following error:

```py
>>> df.show()
org.apache.spark.api.python.PythonException: Traceback (most recent call last):
...
ValueError: not enough values to unpack (expected 2, got 0)
```

### Does this PR introduce _any_ user-facing change?

Yes, Pandas UDF will accept no-column pandas DataFrame.

### How was this patch tested?

Added related tests.

### CODE REVIEW

**Comment by ueshin** [comment]
cc @HyukjinKwon @allisonwang-db

**Review by HyukjinKwon — APPROVED** [review]

File: python/pyspark/sql/pandas/serializers.py [hunk 1]
```diff
@@ -385,37 +385,28 @@ def _create_struct_array(self, df, arrow_struct_type):
         """
         import pyarrow as pa
 
-        # Input partition and result pandas.DataFrame empty, make empty Arrays with struct
-        if len(df) == 0 and len(df.columns) == 0:
-            arrs_names = [
-                (pa.array([], type=field.type), field.name) for field in arrow_struct_type
-            ]
+        if len(df.columns) == 0:
+            return pa.array([{}] * len(df), arrow_struct_type)
```
Comment (allisonwang-db) [inline python/pyspark/sql/pandas/serializers.py:389]: Under what scenario would it be possible for a DataFrame to have rows but no columns? i.e len(df.columns) == 0 but len(df) > 0?

File: python/pyspark/sql/pandas/serializers.py [hunk 2]
```diff
@@ -385,37 +385,28 @@ def _create_struct_array(self, df, arrow_struct_type):
...
         # Assign result columns by schema name if user labeled with strings
         elif self._assign_cols_by_name and any(isinstance(name, str) for name in df.columns):
```
Comment (allisonwang-db) [inline python/pyspark/sql/pandas/serializers.py:391]:
```suggestion
        if self._assign_cols_by_name and any(isinstance(name, str) for name in df.columns):
```

File: python/pyspark/sql/tests/test_udtf.py [hunk 3]
```diff
@@ -487,6 +487,14 @@ def eval(self, x: int):
 
         self.assertEqual(TestUDTF(lit(1)).collect(), [Row(x={1: "1"})])
 
+    def test_udtf_with_empty_output_types(self):
+        @udtf(returnType=StructType())
+        class TestUDTF:
+            def eval(self):
+                yield tuple()
+
+        self.assertEqual(TestUDTF().collect(), [Row()])
```
Comment (allisonwang-db) [inline python/pyspark/sql/tests/test_udtf.py:496]: Let's use assertDataFrameEqual :)

**Review by allisonwang-db — APPROVED** [review]
Thanks for fixing this!

File: python/pyspark/sql/pandas/serializers.py (hunk 1 above)
Comment (ueshin) [reply python/pyspark/sql/pandas/serializers.py:389]: For example:

```py
>>> pdf = pd.DataFrame(index=[1,2,3])
>>> pdf
Empty DataFrame
Columns: []
Index: [1, 2, 3]
>>> len(pdf)
3
>>> len(pdf.columns)
0
```

File: python/pyspark/sql/tests/test_udtf.py [hunk 4]
```diff
@@ -932,8 +940,7 @@ def eval(self, *args):
                     StructType().add("col0", IntegerType()).add("col1", StringType()),
                     [Row(a=1, b="x")],
                 ),
-                # TODO(SPARK-44479): Support Python UDTFs with empty schema
-                # (func(), StructType(), [Row()]),
+                (func(), StructType(), [Row()]),
```
Comment (ueshin) [inline python/pyspark/sql/tests/test_udtf.py:943]: We should backport this PR except for this line.

File: python/pyspark/sql/connect/types.py [hunk 5]
```diff
@@ -170,14 +170,16 @@ def pyspark_types_to_proto_types(data_type: DataType) -> pb2.DataType:
...
         for field in data_type.fields:
             struct_field = pb2.DataType.StructField()
             struct_field.name = field.name
             struct_field.data_type.CopyFrom(pyspark_types_to_proto_types(field.dataType))
             struct_field.nullable = field.nullable
             if field.metadata is not None and len(field.metadata) > 0:
                 struct_field.metadata = json.dumps(field.metadata)
-            ret.struct.fields.append(struct_field)
+            struct.fields.append(struct_field)
+        ret.struct.CopyFrom(struct)
```
Comment (ueshin) [inline python/pyspark/sql/connect/types.py:182]: We might want to backport this part to 3.4, too.

**Review by allisonwang-db — APPROVED** [review]

File: python/pyspark/sql/connect/types.py (hunk 5 above)
Comment (allisonwang-db) [reply python/pyspark/sql/connect/types.py:182]: Good catch!

File: python/pyspark/sql/tests/test_udtf.py (hunk 4 above)
Comment (ueshin) [reply python/pyspark/sql/tests/test_udtf.py:943]: https://github.com/apache/spark/pull/42176

**Comment by ueshin** [comment]
Thanks! merging to master.

**Pull request closed without merging**